lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1413,6 +1413,48 @@ RequirementSignatureRequest::evaluate(Evaluator &evaluator, ProtocolDecl *proto)
   return reqSignature->getRequirements();
 }
 
+llvm::Expected<Type>
+DefaultDefinitionTypeRequest::evaluate(Evaluator &evaluator,
+                                       AssociatedTypeDecl *assocType) const {
+  if (assocType->Resolver) {
+    auto defaultType = assocType->Resolver->loadAssociatedTypeDefault(
+                                    assocType, assocType->ResolverContextData);
+    assocType->Resolver = nullptr;
+    assocType->DefaultDefinition.setType(defaultType);
+    return defaultType;
+  }
+
+  auto &context = assocType->getASTContext();
+  auto &TC = TypeChecker::createForContext(assocType->getASTContext());
+
+  TypeLoc &defaultDefinition = assocType->DefaultDefinition;
+  if (!defaultDefinition.isNull()) {
+    if (TC.validateType(defaultDefinition,
+                        TypeResolution::forInterface(
+                                                  assocType->getDeclContext()),
+                        None)) {
+      defaultDefinition.setInvalidType(context);
+    } else {
+      // associatedtype X = X is invalid
```

Comment:
Can you move the diagnostic check to DeclChecker::visitAssociatedTypeDecl() in TypeCheckDecl.cpp? There's no reason to do it for non-primary files